### PR TITLE
gh-120376: operator.in_ & operator.not_in

### PR DESCRIPTION
**Pull request by dg-pb**
* Issue: gh-120376

----
📚 Documentation preview 📚: https://cpython-previews--120379.org.readthedocs.build/

### CODE REVIEW

File: Lib/operator.py [hunk 1]
```diff
@@ -10,7 +10,8 @@
 This is the pure Python implementation of the module.
 """
 
-__all__ = ['abs', 'add', 'and_', 'attrgetter', 'call', 'concat', 'contains', 'countOf',
+__all__ = ['abs', 'add', 'and_', 'attrgetter', 'call', 'concat',
+           'in_', 'not_in', 'contains', 'countOf',
```
Comment (picnixz) [inline Lib/operator.py]: Please keep the alphabetical order of export even though it might not make sense.

File: Lib/operator.py [hunk 2]
```diff
@@ -150,10 +151,22 @@ def concat(a, b):
         raise TypeError(msg)
     return a + b
 
+
+def in_(a, b):
+    "Same as a in b."
+    return a in b
+
```
Comment (picnixz) [inline Lib/operator.py:156]:
```suggestion
```

File: Lib/operator.py [hunk 3]
```diff
@@ -150,10 +151,22 @@ def concat(a, b):
         raise TypeError(msg)
     return a + b
 
+
```
Comment (picnixz) [inline Lib/operator.py]:
```suggestion
```

File: Lib/operator.py [hunk 4]
```diff
@@ -150,10 +151,22 @@ def concat(a, b):
...
+def in_(a, b):
+    "Same as a in b."
+    return a in b
+
+
+def not_in(a, b):
+    "Same as a not in b."
+    return a not in b
+
```
Comment (picnixz) [inline Lib/operator.py:160]:
```suggestion
```

File: Lib/operator.py [hunk 5]
```diff
@@ -150,10 +151,22 @@ def concat(a, b):
...
+
 def contains(a, b):
     "Same as b in a (note reversed operands)."
     return b in a
 
+
```
Comment (picnixz) [inline Lib/operator.py]:
```suggestion
```

Please do not add new lines if they are not needed. Keep the same style for the whole file.

File: Lib/test/test_operator.py [hunk 6]
```diff
@@ -290,6 +290,22 @@ def test_rshift(self):
         self.assertEqual(operator.rshift(5, 0), 5)
         self.assertRaises(ValueError, operator.rshift, 2, -1)
 
+    def test_in(self):
+        operator = self.module
+        self.assertRaises(TypeError, operator.in_)
+        self.assertRaises(TypeError, operator.in_, None, None)
+        self.assertRaises(ZeroDivisionError, operator.in_, 1, BadIterable())
+        self.assertTrue(operator.in_(2, range(4)))
+        self.assertFalse(operator.in_(5, range(4)))
```
Comment (picnixz) [inline Lib/test/test_operator.py:299]:
```suggestion
        self.assertTrue(operator.in_(2, [1, 2, 3]))
        self.assertFalse(operator.in_(5, [1, 2, 3]))
```

No need to complicate the tests clarity IMO.

File: Lib/test/test_operator.py [hunk 7]
```diff
@@ -290,6 +290,22 @@ def test_rshift(self):
...
+
+    def test_not_in(self):
+        operator = self.module
+        self.assertRaises(TypeError, operator.not_in)
+        self.assertRaises(TypeError, operator.not_in, None, None)
+        self.assertRaises(ZeroDivisionError, operator.not_in, 1, BadIterable())
+        self.assertFalse(operator.not_in(2, range(4)))
+        self.assertTrue(operator.not_in(5, range(4)))
```
Comment (picnixz) [inline Lib/test/test_operator.py:307]: Same observation as above.

File: Misc/NEWS.d/next/Library/2024-06-12-01-12-19.gh-issue-120376.R0lyvC.rst [hunk 8]
```diff
@@ -0,0 +1 @@
+:func:`operator.in_` and :func:`operator.not_in` added to ``operator`` module.
```
Comment (picnixz) [inline Misc/NEWS.d/next/Library/2024-06-12-01-12-19.gh-issue-120376.R0lyvC.rst]:
```suggestion
Add :func:`operator.in_` and :func:`operator.not_in` to the :mod:`operator` module.
```

File: Lib/operator.py (hunk 5 above)
Comment (nineteendo) [inline Lib/operator.py:169]: Could you leave this newline unchanged?

```suggestion
```

File: Modules/_operator.c [hunk 9]
```diff
@@ -474,6 +474,44 @@ _operator_iconcat_impl(PyObject *module, PyObject *a, PyObject *b)
...
+Same as a not in b.
+[clinic start generated code]*/
+
+static int
+_operator_not_in_impl(PyObject *module, PyObject *a, PyObject *b)
+/*[clinic end generated code: output=93a4db26908835bd input=203499d558a29ee3]*/
+{
+    int result = PySequence_Contains(b, a);
+    if (result != -1) {
+        result = 1 - result;
```
Comment (nineteendo) [inline Modules/_operator.c]:
```suggestion
        result = !result;
```

File: Lib/operator.py (hunk 5 above)
Comment (nineteendo) [reply Lib/operator.py]: I thought just the same.

Comment (picnixz) [reply Lib/operator.py]: (Great minds think alike)

File: Modules/_operator.c [hunk 10]
```diff
@@ -474,6 +474,44 @@ _operator_iconcat_impl(PyObject *module, PyObject *a, PyObject *b)
...
+    /
+
+Same as a in b.
+[clinic start generated code]*/
+
+static int
+_operator_in__impl(PyObject *module, PyObject *a, PyObject *b)
+/*[clinic end generated code: output=cd4eca456096fa9a input=8bd20ace23fe9808]*/
+{
+    return PySequence_Contains(b, a);
```
Comment (vstinner) [inline Modules/_operator.c:491]: This code is wrong. You must treat -1 as an error. Same remark for `_operator.not_in()`.

Something like:

```c
int res = PySequence_Contains(b, a);
if (res < 0) {
    return NULL;
}
return PyBool_FromLong(res);
```

Comment (nineteendo) [reply Modules/_operator.c:491]: That's handled by argument clinic like for `_operator_contains()`:

https://github.com/python/cpython/blob/9fcd90d4aad3ceb5de894bc2b643f98ac97048c8/Modules/clinic/_operator.c.h#L914-L918

Comment (vstinner) [reply Modules/_operator.c:491]: Oh, I didn't know! Good to know :-)

File: Lib/test/test_operator.py (hunk 6 above)
Comment (dg-pb) [reply Lib/test/test_operator.py:299]: This is copy+paste of `contains` test. It is simpler if tests are alike. And I don't want to change `test_contains`.

File: Modules/_operator.c [hunk 11]
```diff
@@ -474,6 +474,44 @@ _operator_iconcat_impl(PyObject *module, PyObject *a, PyObject *b)
...
 }
 
+/*[clinic input]
+_operator.in_ -> bool
+
+    a: object
+    b: object
+    /
+
+Same as a in b.
```
Comment (vstinner) [inline Modules/_operator.c:484]:
```suggestion
Same as: a in b.
```

File: Modules/_operator.c [hunk 12]
```diff
@@ -474,6 +474,44 @@ _operator_iconcat_impl(PyObject *module, PyObject *a, PyObject *b)
...
+}
+
+/*[clinic input]
+_operator.not_in -> bool
+
+    a: object
+    b: object
+    /
+
+Same as a not in b.
```
Comment (vstinner) [inline Modules/_operator.c:501]:
```suggestion
Same as: a not in b.
```

**Review by vstinner — APPROVED** [review]
LGTM. I just left a minor suggestion for the docstring.

File: Modules/_operator.c (hunk 11 above)
Comment (dg-pb) [reply Modules/_operator.c:484]: I have had a quick thought about this too. But none of other docstrings have a colon like that and adding it just for these 2 doesn't seem right.

I would probably add colon to all of them, but I don't think this is my decision to make.

**Review by vstinner — CHANGES_REQUESTED** [review]
Haha, in fact, the function already exists: https://docs.python.org/dev/library/operator.html#operator.contains

**Comment by dg-pb** [comment]
> Haha, in fact, the function already exists: https://docs.python.org/dev/library/operator.html#operator.contains

Yes, aware of it. `in_` is a reverse-argument order `__contains__`. `not_in` does not exist.

See, https://discuss.python.org/t/proposed-additions-to-operator/49846/16 for my take on why adding these is potentially be a good idea.

**Comment by dg-pb** [comment]
@vstinner I have laid out my reasoning and examples in discussion. One last thing I can do to support this is to provide use-case search result count:

```
336 - /operator\.contains\(b, a\)/
480 - /lambda a, b\: a in b/
```

And I have laid out my reasoning from personal experience in https://discuss.python.org/t/proposed-additions-to-operator/49846/16 providing a contrast with reversal of other functions.

**Comment by dg-pb** [comment]
I have made the requested changes; please review again

**Comment by nineteendo** [comment]
> I have made the requested changes; please review again

That's not how that works, you haven't changed anything. You can ask to consider the information though.

- \>285 files with [`/lambda \w+, \w+: operator\.contains\(\w+, \w+\)/`](https://github.com/search?q=%2Flambda+%5Cw%2B%2C+%5Cw%2B%3A+operator%5C.contains%5C%28%5Cw%2B%2C+%5Cw%2B%5C%29%2F+lang%3APython&type=code)
- \>96 files with [`/return operator.contains\(\w+, \w+\)/`](https://github.com/search?q=%2Freturn+operator.contains%5C%28%5Cw%2B%2C+%5Cw%2B%5C%29%2F+lang%3APython&type=code)

**Comment by AlexWaygood** [comment]
I agree with Victor and the others who have commented on the Discourse thread that it's not worth it to add a function that is identical to `operator.contains()` except for the reversed order of the arguments. It's not a goal for the `operator` module to be exhaustive over all possible operator-like convenience functions, and nor should it be.

I can see more of an argument for the `not_in` addition.

**Comment by nineteendo** [comment]
Hmm, wouldn't it be better to add `not_contains()` then? Otherwise you expect `in_()` to exist as well.

**Comment by picnixz** [comment]
Personally, I would expect `in` if you give me `contains` so that's why I'm actually supportive of the `in_` and `not_in`. The reason is that I have `is_` and `not_is` so... having the same for the `in` keyword felt natural to me.

**Comment by nineteendo** [comment]
I suggest you close the issue and PR and create new ones for `not_contains()` or `contains_not()`.

**Comment by dg-pb** [comment]
> It's not a goal for the `operator` module to be exhaustive over all possible operator-like convenience functions, and nor should it be.

I agree, my motivation is based on a need and careful consideration.

People are implementing `in_` in their own libraries. >400 implementations with the `exact` search. How many more with different argument names or similar variations - unknown.

And I have explained why it makes sense to me and what sort of functionality it supports.

The fact these somewhat naturally fit in between `contains` and `is`&`not_is` is a supporting argument, not primary motivation.

**Comment by dg-pb** [comment]
I have re-evaluated and incorporated feedback as best as I could.

So I wavered a bit on this one, but came back to my initial position.

What made up my mind was the fact that `in` and `not in` provide default argument order in `ast`:
```python
Compare(
    left=Name(id='a', ctx=Load()),
        ops=[In()],
        comparators=[
        Name(id='b', ctx=Load())]))])
```

So they are generally useful utilities and also important functions for frameworks that build on `ast`.

`not_contains` is not as essential, because it follows argument order that is present in `__contains__`. Which is bound to return a `bool/truth` value, and is automatically inverted when invoked with `not in`. So there is no usefulness argument for it as `__contains__` satisfy both `in` and `not in`.

While `ast` has `In()` and `NotIn()`, which are 2 independent operators.

**Comment by dg-pb** [comment]
I have made the requested changes; please review again

**Comment by nineteendo** [comment]
> I have made the requested changes; please review again

Again: don't say that if you haven't changed anything. You're supposed to "have made the requested changes".

If vstinner agrees with you after posting the new information, he'll drop his stale review.

**Comment by vstinner** [comment]
> @vstinner: please review the changes made to this pull request.

@AlexWaygood and I are against adding `operator.in_()`. Updating the same PR is not going to change my mind.

**Comment by dg-pb** [comment]
> Updating the same PR is not going to change my mind.

I just thought providing new perspectives / arguments and addressing criticism can have an impact.

You kinda left me hanging there, I addressed criticism so I was expecting some response to why my reasoning doesn't cut it. 

And I didn't mean to ping you this way second time, I was just testing how things work as there was still "change requested" and I thought maybe it reappeared because I converted it to draft and then back. Anyway.

Ok, closing this. Thank you for your time considering this.

**Comment by dg-pb** [comment]
> Again: don't say that if you haven't changed anything. You're supposed to "have made the requested changes".
> 
> If vstinner agrees with you after posting the new information, he'll drop his stale review. Otherwise you need to do what he told you to do.

He didn't tell me to do anything.

Generally, I am very confused by your interference. I don't need you explaining me the obvious.

**Comment by nineteendo** [comment]
> He didn't tell me to do anything.

He told you implicitly:

> Haha, in fact, the function already exists: https://docs.python.org/dev/library/operator.html#operator.contains

If the function already exists (with different argument order), that means that it's not needed. (There should be one...)

> I don't need you explaining me the obvious.

I am confused as to why you stated you made changes when you didn't, that's all. I'm totally fine with you arguing about it to try to change our mind.

**Comment by vstinner** [comment]
I'm open to consider adding "not contains", but I'm not sure about the best name for that.

But I don't think that we should add `in_(a,b)` to just exchange argument order compare to `contains(a,b)`.

**Comment by dg-pb** [comment]
I have no interest in `not_contains`, nor do I see its value. I will leave this to someone else then.

**Comment by vstinner** [comment]
> I have no interest in not_contains, nor do I see its value. I will leave this to someone else then.

Ok. Thanks for your contribution anyway. I was close to merge your PR when I saw that `operator.in_()` already exists, sorry about that :-) The PR itself LGTM, it just that we decided that `in_()` and `contains()` are redundant.

**Comment by dg-pb** [comment]
No worries at all, maybe it is for the best.

These 2 functions wouldn't solve the root of what is causing problems to me anyways.

Most of other operators and `object`'s functionality works flexibly and in harmony between `ast` and `__magicks__`, while I am having troubles with containment. Things that force me to write hacks and special cases:
a) forced `bool` return from `__contains__`
b) lack of `__rcontains__`, while `ast.NotIn()` exists
c) revered argument order of `ast.BinOp(In|NotIn)` compared to `__contains__`.

Maybe I will have some better ideas down the line.

**Pull request closed without merging**